File: Difficulty: Hard/Articulation Point - II/articulation-point-ii.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
  vector<int> tin,vis,low,mark;
    int timer = 0;

    void findArticularPoint(int node,int parent,vector<int> adj[]){
        tin[node]=low[node]=timer;
        timer++;
        vis[node] =1;
        
        int child=0;
        for(auto it : adj[node]){
            if(it == parent) continue;
            if(!vis[it]){
                findArticularPoint(it,node,adj);
                low[node]=min(low[node],low[it]);
                
                if(low[it]>=tin[node] && parent != -1) 
                    mark[node]=1;
                
                child++;
            }
            else{
                low[node]=min(tin[it],low[node]);
            }
        }
        if(child>1 && parent==-1) mark[node]=1;
    }
   
    vector<int> articulationPoints(int V, vector<vector<int>>& edges) {
        // Code here
         // Code here
         vector<int>adj[V];
         for(int i=0;i<edges.size();i++)
         {
             int x=edges[i][0];
             int y=edges[i][1];
             adj[x].push_back(y);
             adj[y].push_back(x);
         }
         tin.resize(V);
        vis.resize(V,0);
        low.resize(V);
        mark.resize(V,0);
        for(int i=0;i<V;i++){
            if(vis[i]==0){
                timer=0;
             findArticularPoint(i,-1,adj);
             
            }
        }
        vector<int> ans;
        for(int i=0; i<V; i++)
            if(mark[i]==1) ans.push_back(i);
            
        if(ans.size()==0) return {-1} ;
        return ans;
        
    }
};


//{ Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int V, E;
        cin >> V >> E;
        vector<vector<int>> edges;
        for (int i = 0; i < E; i++) {
            int u, v;
            cin >> u >> v;
            edges.push_back({u, v});
        }
        Solution obj;
        vector<int> ans = obj.articulationPoints(V, edges);
        sort(ans.begin(), ans.end());
        for (auto i : ans)
            cout << i << " ";
        cout << "\n";

        cout << "~"
             << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Hard/Articulation Point - II/articulation-point-ii.java
import java.util.*;

class Solution {

    static int timer;
    static int[] tin, low, vis, mark;

    static void dfs(int node, int parent, ArrayList<ArrayList<Integer>> adj) {

        vis[node] = 1;
        tin[node] = low[node] = timer;
        timer++;

        int child = 0;

        for (int it : adj.get(node)) {

            if (it == parent) {
                continue;
            }

            if (vis[it] == 0) {

                dfs(it, node, adj);

                low[node] = Math.min(low[node], low[it]);

                // articulation condition (non-root)
                if (low[it] >= tin[node] && parent != -1) {
                    mark[node] = 1;
                }

                child++;
            } else {
                // back edge
                low[node] = Math.min(low[node], tin[it]);
            }
        }

        // root condition
        if (parent == -1 && child > 1) {
            mark[node] = 1;
        }
    }

    static ArrayList<Integer> articulationPoints(int V, int[][] edges) {

        // Step 1: Build adjacency list
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();

        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }

        int i = 0;
        while (i < edges.length) {
            int u = edges[i][0];
            int v = edges[i][1];

            adj.get(u).add(v);
            adj.get(v).add(u);

            i++;
        }

        // Step 2: Initialize arrays
        tin = new int[V];
        low = new int[V];
        vis = new int[V];
        mark = new int[V];
        timer = 0;

        // Step 3: DFS for all components
        int j = 0;
        while (j < V) {
            if (vis[j] == 0) {
                dfs(j, -1, adj);
            }
            j++;
        }

        // Step 4: Collect answer
        ArrayList<Integer> ans = new ArrayList<>();

        int k = 0;
        while (k < V) {
            if (mark[k] == 1) {
                ans.add(k);
            }
            k++;
        }

        if (ans.size() == 0) {
            ans.add(-1);
        }

        return ans;
    }
}